src/Interpreters/join_common.cpp

```diff
@@ -164,6 +188,45 @@ Columns materializeColumns(const Block & block, const Names & names)
     return materialized;
 }
 
+bool canCastJoinColumns(const Block & left_block, const TableJoin & table_join)
+{
+    bool is_left_or_inner = isLeft(table_join.kind()) || isInner(table_join.kind());
+    bool key_names_match = table_join.keyNamesLeft().size() == table_join.keyNamesRight().size();
+    return left_block && key_names_match && (!table_join.hasUsing() || is_left_or_inner);
```

Comment:
So it doesn't work for `using` yet? I think `using` is the main use case for this feature -- in `on equality`, you can add casts manually, but in `using` you cannot, so this is where we really need the implicit cast.

Reply:
With `using` type of key column depends on kind of join. For left (right) join it will have type from left (right) table. For `full join` it should be supertype, that is not always possible to get (for example for string <-> int, or big signed and unsigned int).
Now it works with `LEFT JOIN ... USING`, and I can imagine how to implement for `right join`. But what to do with full join?

Reply:
`using key` is just a syntactic sugar for `on left_table.key = right_table.key`, the behavior should be the same.

A value that is not convertible never matches the join condition, so logically it behaves the same way as NULL. All kinds of outer joins work with these non-convertible values the same way they work with NULLs. So I think if you convert the right column to a Nullable type, it should just 'magically' work.

In cases where the left type is a supertype of the right type, the conversion is always possible, so you can omit Nullable.

Reply:
(I'm reading the Postgres docs for `USING` https://www.postgresql.org/docs/current/queries-table-expressions.html)

Reply:
```
select a
from ( select number + 2 as a from system.numbers limit 2) as t1
full join (select number + 1 as a from system.numbers limit 3) as t2
on t1.a == t2.a
ORDER BY a;
```
gives `0 1 2` as if `a` is same as `t1.a`

But
```
select a
from ( select number + 2 as a from system.numbers limit 2) as t1
full join (select number + 1 as a from system.numbers limit 3) as t2
using (a)
ORDER BY a;
```
gives `1 2 3` as if `a` contains part of `t1.a` and part of `t2.a`

Reply:
The tables are 2, 3 and 1, 2, 3, right? In the `on` case, the `a` is interpreted as t1.a, probably that's why you're getting a zero, from the t2 row that doesn't have the match. If you qualify both columns explicitly, it is more clear:
```
*2* :) select l.a, r.a from (select arrayJoin([0, 1]) a) l full join (select arrayJoin([1, 2]) a) r on l.a = r.a;

┌─a─┬─r.a─┐
│ 0 │   0 │
│ 1 │   1 │
└───┴─────┘
┌─a─┬─r.a─┐
│ 0 │   2 │
└───┴─────┘
```

So this is not related to the set of rows returned by the join, but to the fact that `a` can be interpreted as `t1.a` or `t2.a`. Maybe we should fix this difference as well, but it's not related to this PR.

By the way, seems like the unqualified key column is interpreted differently for `using` and `on`:

```
postgres=# select a, l.a, r.a from (select unnest('{0, 1}'::int[]) a) l full join (select unnest('{1, 2}'::int[]) a) r using (a);
 a | a | a 
---+---+---
 0 | 0 |  
 1 | 1 | 1
 2 |   | 2
(3 rows)

postgres=# select a, l.a, r.a from (select unnest('{0, 1}'::int[]) a) l full join (select unnest('{1, 2}'::int[]) a) r on l.a = r.a;
2021-01-13 18:21:50.679 MSK [175080] ERROR:  column reference "a" is ambiguous at character 8
2021-01-13 18:21:50.679 MSK [175080] STATEMENT:  select a, l.a, r.a from (select unnest('{0, 1}'::int[]) a) l full join (select unnest('{1, 2}'::int[]) a) r on l.a = r.a;
ERROR:  column reference "a" is ambiguous
LINE 1: select a, l.a, r.a from (select unnest('{0, 1}'::int[]) a) l...
               ^
```

But probably we don't have to do anything about this now, it's not related to type conversion.

Reply:
Ah, I think I understand what you mean now, the behavior for `using` should be like in Postgres so we need a supertype for key columns, and we need to convert both tables. But it's broken now anyway? Maybe we can fix it later...

Reply:
Now it is works in following way: for LEFT join type of key column is taken from left table, for RIGHT it's taken from right table, FULL join won't work if cast is required (throws usual error about type mismatch).